File: prompt-ide/src/lib/backend.ts
import { getLocalServerUrl } from './types';

const TOKEN_KEY = 'inkwell-jwt-token';

// --- Token management ---

export function getToken(): string | null {
  return localStorage.getItem(TOKEN_KEY);
}

export function setToken(token: string): void {
  localStorage.setItem(TOKEN_KEY, token);
}

export function clearToken(): void {
  localStorage.removeItem(TOKEN_KEY);
}

function getBaseUrl(): string {
  return getLocalServerUrl().replace(/\/$/, '');
}

// --- HTTP helper ---

async function request<T>(method: string, path: string, body?: unknown): Promise<T> {
  const headers: Record<string, string> = { 'Content-Type': 'application/json' };
  const token = getToken();
  if (token) headers['Authorization'] = `Bearer ${token}`;

  const res = await fetch(`${getBaseUrl()}/api${path}`, {
    method,
    headers,
    body: body ? JSON.stringify(body) : undefined,
  });

  if (res.status === 401) {
    clearToken();
    throw new Error('UNAUTHORIZED');
  }

  if (!res.ok) {
    const text = await res.text();
    throw new Error(text || `HTTP ${res.status}`);
  }

  if (res.status === 204) return undefined as T;
  return res.json();
}

// --- Auth ---

export interface BackendUser {
  id: string;
  email: string;
  display_name: string;
  avatar: string;
}

interface AuthResponse {
  token: string;
  user: BackendUser;
}

export async function register(email: string, password: string, displayName: string): Promise<AuthResponse> {
  const resp = await request<AuthResponse>('POST', '/auth/register', { email, password, display_name: displayName });
  setToken(resp.token);
  return resp;
}

export async function login(email: string, password: string): Promise<AuthResponse> {
  const resp = await request<AuthResponse>('POST', '/auth/login', { email, password });
  setToken(resp.token);
  return resp;
}

export async function getMe(): Promise<BackendUser> {
  return request<BackendUser>('GET', '/auth/me');
}

export function logout(): void {
  clearToken();
}

export function isLoggedIn(): boolean {
  return !!getToken();
}

// --- OAuth ---

export async function oauthGoogle(token: string): Promise<AuthResponse> {
  const resp = await request<AuthResponse>('POST', '/auth/oauth/google', { token });
  setToken(resp.token);
  return resp;
}

export async function oauthGithub(code: string): Promise<AuthResponse> {
  const resp = await request<AuthResponse>('POST', '/auth/oauth/github', { code });
  setToken(resp.token);
  return resp;
}

// --- Presence ---

export interface PresenceUser {
  user_id: string;
  display_name: string;
}

export async function setPresence(projectId: string): Promise<void> {
  return request('POST', '/presence', { project_id: projectId });
}

export async function getPresence(projectId: string): Promise<PresenceUser[]> {
  return request('GET', `/presence/${projectId}`);
}

// --- Workspaces ---

export interface BackendWorkspace {
  id: string;
  name: string;
  description: string;
  color: string;
  user_id: string;
  created_at: number;
  updated_at: number;
}

export async function listWorkspaces(): Promise<BackendWorkspace[]> {
  return request('GET', '/workspaces');
}

export async function createWorkspace(data: { name: string; color: string; description?: string }): Promise<BackendWorkspace> {
  return request('POST', '/workspaces', data);
}

export async function updateWorkspace(id: string, data: { name?: string; color?: string; description?: string }): Promise<BackendWorkspace> {
  return request('PUT', `/workspaces/${id}`, data);
}

export async function deleteWorkspace(id: string): Promise<void> {
  return request('DELETE', `/workspaces/${id}`);
}

// --- Projects ---

export interface BackendProject {
  id: string;
  name: string;
  user_id: string;
  workspace_id: string | null;
  blocks_json: string;
  variables_json: string;
  framework: string | null;
  tags_json: string;
  created_at: number;
  updated_at: number;
}

export async function listProjects(): Promise<BackendProject[]> {
  return request('GET', '/projects');
}

export async function createProject(data: {
  id?: string; name: string; workspace_id?: string | null;
  blocks_json: string; variables_json?: string; framework?: string | null; tags_json?: string;
}): Promise<BackendProject> {
  return request('POST', '/projects', data);
}

export async function updateProject(id: string, data: {
  name?: string; workspace_id?: string | null;
  blocks_json?: string; variables_json?: string; framework?: string | null; tags_json?: string;
}): Promise<BackendProject> {
  return request('PUT', `/projects/${id}`, data);
}

export async function deleteProject(id: string): Promise<void> {
  return request('DELETE', `/projects/${id}`);
}

// --- Versions ---

export interface BackendVersion {
  id: string;
  project_id: string;
  blocks_json: string;
  variables_json: string;
  label: string;
  created_at: number;
}

export async function listVersions(projectId: string): Promise<BackendVersion[]> {
  return request('GET', `/projects/${projectId}/versions`);
}

export async function createVersion(projectId: string, data: { blocks_json: string; variables_json: string; label: string }): Promise<BackendVersion> {
  return request('POST', `/projects/${projectId}/versions`, data);
}

// --- Executions ---

export interface BackendExecution {
  id: string;
  project_id: string;
  model: string;
  provider: string;
  prompt: string;
  response: string;
  tokens_in: number;
  tokens_out: number;
  cost: number;
  latency_ms: number;
  created_at: number;
}

export async function listExecutions(projectId: string): Promise<BackendExecution[]> {
  return request('GET', `/projects/${projectId}/executions`);
}

export async function createExecution(projectId: string, data: {
  model: string; provider: string; prompt: string; response: string;
  tokens_in: number; tokens_out: number; cost: number; latency_ms: number;
}): Promise<BackendExecution> {
  return request('POST', `/projects/${projectId}/executions`, data);
}

// --- Frameworks ---

export interface BackendFramework {
  id: string;
  name: string;
  description: string;
  blocks_json: string;
  user_id: string;
  created_at: number;
  updated_at: number;
}

export async function listFrameworks(): Promise<BackendFramework[]> {
  return request('GET', '/frameworks');
}

export async function createFramework(data: { name: string; description?: string; blocks_json: string }): Promise<BackendFramework> {
  return request('POST', '/frameworks', data);
}

export async function updateFramework(id: string, data: { name?: string; description?: string; blocks_json?: string }): Promise<BackendFramework> {
  return request('PUT', `/frameworks/${id}`, data);
}

export async function deleteFramework(id: string): Promise<void> {
  return request('DELETE', `/frameworks/${id}`);
}

// --- Config ---

export async function getConfig(): Promise<Record<string, string>> {
  return request('GET', '/config');
}

export async function setConfig(config: Record<string, string>): Promise<void> {
  return request('PUT', '/config', { config });
}

// --- GPU Nodes (Fleet) ---

export interface GpuNode {
  id: string;
  user_id: string;
  name: string;
  hostname: string;
  gpu_info: string;
  last_heartbeat: number;
  status: 'online' | 'offline' | 'degraded';
  capabilities_json: string;
  address: string;
  created_at: number;
}

export interface GpuNodeCapabilities {
  stt: {
    model_loaded: boolean;
    active_model?: string;
    available_models: string[];
  };
  llm: {
    ollama_connected: boolean;
    models: { name: string; size_gb: number; parameter_size?: string }[];
  };
}

export function parseCapabilities(node: GpuNode): GpuNodeCapabilities {
  try {
    return JSON.parse(node.capabilities_json);
  } catch {
    return { stt: { model_loaded: false, available_models: [] }, llm: { ollama_connected: false, models: [] } };
  }
}

export async function listNodes(): Promise<GpuNode[]> {
  return request('GET', '/nodes');
}

export async function deleteNode(id: string): Promise<void> {
  return request('DELETE', `/nodes/${id}`);
}

export async function updateNodeName(id: string, name: string): Promise<void> {
  return request('PUT', `/nodes/${id}`, { name });
}

export async function routeNode(capability: string, model?: string): Promise<GpuNode> {
  const params = new URLSearchParams({ capability });
  if (model) params.set('model', model);
  return request('GET', `/nodes/route?${params}`);
}

// Helper: find best node for STT from a list
export function findSttNode(nodes: GpuNode[]): GpuNode | null {
  return nodes.find(n => {
    if (n.status !== 'online') return false;
    const caps = parseCapabilities(n);
    return caps.stt.model_loaded;
  }) ?? null;
}

// Helper: find best node for a specific LLM model
export function findLlmNode(nodes: GpuNode[], modelName?: string): GpuNode | null {
  return nodes.find(n => {
    if (n.status !== 'online') return false;
    const caps = parseCapabilities(n);
    if (!caps.llm.ollama_connected) return false;
    if (modelName) return caps.llm.models.some(m => m.name.includes(modelName));
    return true;
  }) ?? null;
}

File: prompt-ide-desktop/src/services/backend.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct BackendClient {
    base_url: String,
    token: Option<String>,
    client: reqwest::Client,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BackendUser {
    pub id: String,
    pub email: String,
    pub display_name: String,
    pub avatar: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuthResponse {
    pub token: String,
    pub user: BackendUser,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BackendWorkspace {
    pub id: String,
    pub name: String,
    pub description: String,
    pub color: String,
    pub user_id: String,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BackendProject {
    pub id: String,
    pub name: String,
    pub user_id: String,
    pub workspace_id: Option<String>,
    pub blocks_json: String,
    pub variables_json: String,
    pub framework: Option<String>,
    pub tags_json: String,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BackendVersion {
    pub id: String,
    pub project_id: String,
    pub blocks_json: String,
    pub variables_json: String,
    pub label: String,
    pub created_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BackendExecution {
    pub id: String,
    pub project_id: String,
    pub model: String,
    pub provider: String,
    pub prompt: String,
    pub response: String,
    pub tokens_in: i64,
    pub tokens_out: i64,
    pub cost: f64,
    pub latency_ms: i64,
    pub created_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BackendFramework {
    pub id: String,
    pub name: String,
    pub description: String,
    pub blocks_json: String,
    pub user_id: String,
    pub created_at: i64,
    pub updated_at: i64,
}

impl BackendClient {
    pub fn new(base_url: &str) -> Self {
        Self {
            base_url: base_url.trim_end_matches('/').to_string(),
            token: None,
            client: reqwest::Client::new(),
        }
    }

    pub fn set_token(&mut self, token: String) {
        self.token = Some(token);
    }

    pub fn clear_token(&mut self) {
        self.token = None;
    }

    pub fn is_logged_in(&self) -> bool {
        self.token.is_some()
    }

    pub fn set_base_url(&mut self, url: &str) {
        self.base_url = url.trim_end_matches('/').to_string();
    }

    fn auth_header(&self) -> Option<String> {
        self.token.as_ref().map(|t| format!("Bearer {t}"))
    }

    async fn get<T: serde::de::DeserializeOwned>(&self, path: &str) -> Result<T, String> {
        let mut req = self.client.get(format!("{}/api{path}", self.base_url));
        if let Some(auth) = self.auth_header() { req = req.header("Authorization", auth); }
        let res = req.send().await.map_err(|e| e.to_string())?;
        if !res.status().is_success() {
            let text = res.text().await.unwrap_or_default();
            return Err(text);
        }
        res.json().await.map_err(|e| e.to_string())
    }

    async fn post<T: serde::de::DeserializeOwned, B: Serialize>(&self, path: &str, body: &B) -> Result<T, String> {
        let mut req = self.client.post(format!("{}/api{path}", self.base_url)).json(body);
        if let Some(auth) = self.auth_header() { req = req.header("Authorization", auth); }
        let res = req.send().await.map_err(|e| e.to_string())?;
        if !res.status().is_success() {
            let text = res.text().await.unwrap_or_default();
            return Err(text);
        }
        res.json().await.map_err(|e| e.to_string())
    }

    async fn put<T: serde::de::DeserializeOwned, B: Serialize>(&self, path: &str, body: &B) -> Result<T, String> {
        let mut req = self.client.put(format!("{}/api{path}", self.base_url)).json(body);
        if let Some(auth) = self.auth_header() { req = req.header("Authorization", auth); }
        let res = req.send().await.map_err(|e| e.to_string())?;
        if !res.status().is_success() {
            let text = res.text().await.unwrap_or_default();
            return Err(text);
        }
        res.json().await.map_err(|e| e.to_string())
    }

    async fn delete_req(&self, path: &str) -> Result<(), String> {
        let mut req = self.client.delete(format!("{}/api{path}", self.base_url));
        if let Some(auth) = self.auth_header() { req = req.header("Authorization", auth); }
        let res = req.send().await.map_err(|e| e.to_string())?;
        if !res.status().is_success() {
            let text = res.text().await.unwrap_or_default();
            return Err(text);
        }
        Ok(())
    }

    // --- Auth ---

    pub async fn register(&self, email: &str, password: &str, display_name: &str) -> Result<AuthResponse, String> {
        self.post("/auth/register", &serde_json::json!({ "email": email, "password": password, "display_name": display_name })).await
    }

    pub async fn login(&self, email: &str, password: &str) -> Result<AuthResponse, String> {
        self.post("/auth/login", &serde_json::json!({ "email": email, "password": password })).await
    }

    pub async fn get_me(&self) -> Result<BackendUser, String> {
        self.get("/auth/me").await
    }

    // --- Workspaces ---

    pub async fn list_workspaces(&self) -> Result<Vec<BackendWorkspace>, String> {
        self.get("/workspaces").await
    }

    pub async fn create_workspace(&self, name: &str, color: &str) -> Result<BackendWorkspace, String> {
        self.post("/workspaces", &serde_json::json!({ "name": name, "color": color })).await
    }

    pub async fn delete_workspace(&self, id: &str) -> Result<(), String> {
        self.delete_req(&format!("/workspaces/{id}")).await
    }

    // --- Projects ---

    pub async fn list_projects(&self) -> Result<Vec<BackendProject>, String> {
        self.get("/projects").await
    }

    pub async fn create_project(&self, data: &serde_json::Value) -> Result<BackendProject, String> {
        self.post("/projects", data).await
    }

    pub async fn update_project(&self, id: &str, data: &serde_json::Value) -> Result<BackendProject, String> {
        self.put(&format!("/projects/{id}"), data).await
    }

    pub async fn delete_project(&self, id: &str) -> Result<(), String> {
        self.delete_req(&format!("/projects/{id}")).await
    }

    // --- Versions ---

    pub async fn list_versions(&self, project_id: &str) -> Result<Vec<BackendVersion>, String> {
        self.get(&format!("/projects/{project_id}/versions")).await
    }

    pub async fn create_version(&self, project_id: &str, data: &serde_json::Value) -> Result<BackendVersion, String> {
        self.post(&format!("/projects/{project_id}/versions"), data).await
    }

    // --- Executions ---

    pub async fn list_executions(&self, project_id: &str) -> Result<Vec<BackendExecution>, String> {
        self.get(&format!("/projects/{project_id}/executions")).await
    }

    pub async fn create_execution(&self, project_id: &str, data: &serde_json::Value) -> Result<BackendExecution, String> {
        self.post(&format!("/projects/{project_id}/executions"), data).await
    }

    // --- Frameworks ---

    pub async fn list_frameworks(&self) -> Result<Vec<BackendFramework>, String> {
        self.get("/frameworks").await
    }

    pub async fn create_framework(&self, data: &serde_json::Value) -> Result<BackendFramework, String> {
        self.post("/frameworks", data).await
    }

    pub async fn update_framework(&self, id: &str, data: &serde_json::Value) -> Result<BackendFramework, String> {
        self.put(&format!("/frameworks/{id}"), data).await
    }

    pub async fn delete_framework(&self, id: &str) -> Result<(), String> {
        self.delete_req(&format!("/frameworks/{id}")).await
    }

    // --- Config ---

    pub async fn get_config(&self) -> Result<HashMap<String, String>, String> {
        self.get("/config").await
    }

    pub async fn set_config(&self, config: &HashMap<String, String>) -> Result<(), String> {
        let mut req = self.client.put(format!("{}/api/config", self.base_url))
            .json(&serde_json::json!({ "config": config }));
        if let Some(auth) = self.auth_header() { req = req.header("Authorization", auth); }
        let res = req.send().await.map_err(|e| e.to_string())?;
        if !res.status().is_success() { return Err(res.text().await.unwrap_or_default()); }
        Ok(())
    }
}
